File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.ts
function countMaxOrSubsets(nums: number[]): number {
    const max = nums.reduce((acc, cur)=>acc | cur, 0)
    const dp = new Array(max + 1).fill(0)
    dp[0] = 1
    for(const n of nums) {
        const prev = [...dp]
        for(let i = 0; i <= max; i++) {
            dp[i | n] += prev[i]
        }
    }
    return dp[max]
};

File: 2170-count-number-of-maximum-bitwise-or-subsets/count-number-of-maximum-bitwise-or-subsets.rs
use std::collections::HashMap;

impl Solution {
    pub fn count_max_or_subsets(nums: Vec<i32>) -> i32 {
        let mut hashmap: HashMap<i32, i32> = HashMap::new();
        let mut max_value = 0;
        let mut cur_value = 0;
        hashmap.insert(0, 1);
        for n in nums.iter() {
            max_value |= n;
        }

        for n in nums.iter() {
            for i in (0..=cur_value).rev() {
                let prev = *hashmap.get(&i).unwrap_or(&0);
                *hashmap.entry(i | n).or_insert(0) += prev;
            }
            cur_value |= n;
        }
        *hashmap.get(&max_value).unwrap_or(&0)
    }
}

/*
3 = 1
2 = 1
1 = 1
5 = 1


*/